File: orchard_guidance/perception.py
from __future__ import annotations

from typing import Optional

import cv2
import numpy as np

from .config import GuidanceConfig
from .geometry import camera_points_to_ground, ground_to_camera
from .types import Frame, LineXZ, RowPerception, Trunk


def _fit_line_xz(xs: np.ndarray, zs: np.ndarray, thresh: float) -> Optional[LineXZ]:
    xs = np.asarray(xs, dtype=np.float64)
    zs = np.asarray(zs, dtype=np.float64)
    if xs.size < 2:
        return None
    mask = np.ones(xs.size, dtype=bool)
    intercept = slope = 0.0
    for _ in range(4):
        if int(mask.sum()) < 2:
            return None
        a = np.stack([np.ones(int(mask.sum())), zs[mask]], axis=1)
        coef, *_ = np.linalg.lstsq(a, xs[mask], rcond=None)
        intercept, slope = float(coef[0]), float(coef[1])
        resid = np.abs(intercept + slope * zs - xs)
        mask = resid < thresh
    if int(mask.sum()) < 2:
        return None
    return LineXZ(intercept=intercept, slope=slope)


def _cluster_xz(xs, zs, eps, min_pts):
    n = xs.size
    if n == 0:
        return []
    used = np.zeros(n, dtype=bool)
    trunks = []
    for i in np.argsort(zs):
        if used[i]:
            continue
        dist2 = (xs - xs[i]) ** 2 + (zs - zs[i]) ** 2
        members = dist2 < eps * eps
        count = int(members.sum())
        if count < min_pts:
            continue
        used[members] = True
        trunks.append((float(xs[members].mean()), float(zs[members].mean()), count))
    trunks.sort(key=lambda t: t[1])
    return trunks


def _intersect_image_lines(l0, l1):
    a0, b0, c0 = l0
    a1, b1, c1 = l1
    det = a0 * b1 - a1 * b0
    if abs(det) < 1e-8:
        return None
    x = (b0 * c1 - b1 * c0) / det
    y = (c0 * a1 - c1 * a0) / det
    return float(x), float(y)


def _line_from_two_points(u0, v0, u1, v1):
    return (v0 - v1, u1 - u0, u0 * v1 - u1 * v0)


def _vanishing_from_rgb(rgb, cx):
    gray = cv2.cvtColor(rgb, cv2.COLOR_BGR2GRAY)
    gray = cv2.GaussianBlur(gray, (5, 5), 0)
    edges = cv2.Canny(gray, 60, 160)
    h, w = gray.shape
    lines = cv2.HoughLinesP(edges, 1, np.pi / 180.0, threshold=40, minLineLength=int(h * 0.18), maxLineGap=25)
    if lines is None:
        return None
    segs = lines.reshape(-1, 4)
    left_pts, right_pts = [], []
    for x1, y1, x2, y2 in segs:
        if y1 > y2:
            x1, y1, x2, y2 = x2, y2, x1, y1
        dx, dy = x2 - x1, y2 - y1
        if float(np.hypot(dx, dy)) < 30:
            continue
        angle = abs(np.degrees(np.arctan2(dy, dx)))
        if angle < 18 or angle > 78:
            continue
        mid_x = 0.5 * (x1 + x2)
        if mid_x < cx and x2 < x1:
            left_pts.append((x1, y1, x2, y2))
        elif mid_x > cx and x2 > x1:
            right_pts.append((x1, y1, x2, y2))
        elif mid_x < cx and x2 > x1:
            left_pts.append((x1, y1, x2, y2))
        elif mid_x > cx and x2 < x1:
            right_pts.append((x1, y1, x2, y2))
    if len(left_pts) < 1 or len(right_pts) < 1:
        return None

    def consensus(segs):
        arr = np.array(segs, dtype=np.float64)
        u0, v0 = float(arr[:, 0].mean()), float(arr[:, 1].mean())
        u1, v1 = float(arr[:, 2].mean()), float(arr[:, 3].mean())
        if abs(v1 - v0) < 1:
            return None
        return _line_from_two_points(u0, v0, u1, v1)

    l_line, r_line = consensus(left_pts), consensus(right_pts)
    if l_line is None or r_line is None:
        return None
    return _intersect_image_lines(l_line, r_line)


def _project_line(line, z0, z1, height_agl, frame, config):
    xs = np.array([line.x_at(z0), line.x_at(z1)], dtype=np.float32)
    hs = np.array([height_agl, height_agl], dtype=np.float32)
    zs = np.array([z0, z1], dtype=np.float32)
    cx, cy, cz = ground_to_camera(xs, hs, zs, config.camera_pitch_rad, config.camera_height_m)
    if np.any(cz < 0.3):
        return None
    u, v = frame.intrinsics.cam_to_pixel(cx, cy, cz)
    if not np.isfinite(u).all() or not np.isfinite(v).all():
        return None
    return (float(u[0]), float(v[0])), (float(u[1]), float(v[1]))


def _inner_envelope(xs, zs, side, z_bin=0.9, min_pts=8):
    if xs.size == 0:
        return np.array([]), np.array([])
    z0, z1 = float(zs.min()), float(zs.max())
    bins = np.arange(z0, z1 + z_bin, z_bin)
    ox, oz = [], []
    for b0, b1 in zip(bins[:-1], bins[1:]):
        m = (zs >= b0) & (zs < b1)
        if int(m.sum()) < min_pts:
            continue
        ox.append(float(xs[m].max() if side == "left" else xs[m].min()))
        oz.append(float(0.5 * (b0 + b1)))
    return np.asarray(ox, dtype=np.float64), np.asarray(oz, dtype=np.float64)


def _plausible_vp(vp, width, height):
    u, v = vp
    return (0.18 * width) < u < (0.82 * width) and (-0.15 * height) < v < (0.62 * height)


def _cluster_pens(xs, ys, zs, eps, min_pts):
    n = xs.size
    if n == 0:
        return []
    used = np.zeros(n, dtype=bool)
    out = []
    for i in np.argsort(zs):
        if used[i]:
            continue
        dist2 = (xs - xs[i]) ** 2 + (zs - zs[i]) ** 2
        members = dist2 < eps * eps
        count = int(members.sum())
        if count < min_pts:
            continue
        used[members] = True
        mx, my, mz = float(xs[members].mean()), float(ys[members].mean()), float(zs[members].mean())
        xz_span = float(np.hypot(xs[members].max() - xs[members].min(), zs[members].max() - zs[members].min()))
        y_span = float(ys[members].max() - ys[members].min())
        out.append((mx, my, mz, count, xz_span, y_span))
    return out


def perceive_pens(frame: Frame, config: GuidanceConfig) -> RowPerception:
    """Lock onto two compact nearby objects (pens) and steer between them."""
    result = RowPerception()
    depth = frame.depth_m
    k = frame.intrinsics
    z0, z1 = config.desk_depth_min_m, config.desk_depth_max_m
    valid = (depth > z0) & (depth < z1)
    if int(valid.sum()) < 40:
        result.notes.append("too_few_depth_pixels")
        return result

    vv, uu = np.indices(depth.shape)
    z = depth[valid]
    x, y_down, z = k.pixel_to_cam(uu[valid].astype(np.float32), vv[valid].astype(np.float32), z)
    band = np.abs(y_down) <= config.desk_y_band_m
    x, y_down, z = x[band], y_down[band], z[band]
    if x.size < 20:
        result.notes.append("no_band")
        return result

    clusters = []
    for mx, my, mz, n, xz_span, y_span in _cluster_pens(
        x, y_down, z, config.desk_cluster_eps_m, config.desk_cluster_min_points
    ):
        if xz_span > config.desk_max_span_m:
            continue
        if y_span < config.desk_min_vertical_m and n < config.desk_cluster_min_points * 3:
            continue
        clusters.append((mx, my, mz, n, xz_span, y_span))

    left = [c for c in clusters if c[0] < -0.02]
    right = [c for c in clusters if c[0] > 0.02]
    left.sort(key=lambda c: c[2])
    right.sort(key=lambda c: c[2])
    if not left:
        result.notes.append("left_missing")
    if not right:
        result.notes.append("right_missing")
    if not left or not right:
        result.notes.append("need_both")
        return result

    lx, ly, lz, ln, lspan, _lyspan = left[0]
    rx, ry, rz, rn, rspan, _ryspan = right[0]
    gap = rx - lx
    if gap < config.desk_min_gap_m or gap > config.desk_max_gap_m:
        result.notes.append("gap_rejected")
        return result

    center_x = 0.5 * (lx + rx)
    lat = -center_x
    if abs(lat) > config.desk_max_lat_m:
        result.notes.append("lat_rejected")
        return result

    result.left_line = LineXZ(intercept=lx, slope=0.0)
    result.right_line = LineXZ(intercept=rx, slope=0.0)
    result.centerline = LineXZ(intercept=center_x, slope=0.0)
    result.lateral_error_m = float(lat)
    result.heading_error_rad = 0.0
    result.row_width_m = float(gap)
    result.depth_range_m = float(max(lz, rz))
    lu, lv = k.cam_to_pixel(np.array([lx]), np.array([ly]), np.array([lz]))
    ru, rv = k.cam_to_pixel(np.array([rx]), np.array([ry]), np.array([rz]))
    result.trunks = [
        Trunk(side="left", x_m=float(lx), z_m=float(lz), u=float(lu[0]), v=float(lv[0]), n_points=int(ln)),
        Trunk(side="right", x_m=float(rx), z_m=float(rz), u=float(ru[0]), v=float(rv[0]), n_points=int(rn)),
    ]
    result.confidence = _pen_confidence(ln, rn, lz, rz, lspan, rspan, gap, config)
    return result


def _pen_confidence(ln, rn, lz, rz, lspan, rspan, gap, config: GuidanceConfig) -> float:
    """How clean the two-pen lock is (not how centered you are)."""

    def unit(x: float) -> float:
        return float(np.clip(x, 0.0, 1.0))

    points = unit(min(ln, rn) / 60.0)
    depth_match = unit(1.0 - abs(lz - rz) / 0.30)
    compact = unit(1.0 - max(lspan, rspan) / max(config.desk_max_span_m, 1e-3))
    gap_q = unit(1.0 - abs(gap - 0.30) / 0.40)
    return float(np.clip(0.30 + 0.28 * points + 0.22 * depth_match + 0.12 * compact + 0.08 * gap_q, 0.05, 0.99))


def perceive(frame: Frame, config: GuidanceConfig) -> RowPerception:
    if config.desk_mode:
        return perceive_pens(frame, config)
    result = RowPerception()
    depth = frame.depth_m
    k = frame.intrinsics
    valid = (depth > config.depth_min_m) & (depth < config.depth_max_m)
    if int(valid.sum()) < 80:
        result.notes.append("too_few_depth_pixels")
        return result

    vv, uu = np.indices(depth.shape)
    z = depth[valid]
    x, y_down, z = k.pixel_to_cam(uu[valid].astype(np.float32), vv[valid].astype(np.float32), z)
    gx, agl, gz = camera_points_to_ground(x, y_down, z, config.camera_pitch_rad, config.camera_height_m)
    band = (
        (agl >= config.trunk_height_min_m)
        & (agl <= config.trunk_height_max_m)
        & (gz >= config.depth_min_m)
        & (gz <= config.depth_max_m)
    )
    gx, gz = gx[band], gz[band]
    if gx.size < 40:
        result.notes.append("no_trunk_band")

    left_mask = gx < -config.center_keepout_m
    right_mask = gx > config.center_keepout_m
    left_env_x, left_env_z = _inner_envelope(gx[left_mask], gz[left_mask], "left")
    right_env_x, right_env_z = _inner_envelope(gx[right_mask], gz[right_mask], "right")

    if left_env_x.size >= config.min_trunks_for_line:
        result.left_line = _fit_line_xz(left_env_x, left_env_z, max(config.ransac_thresh_m, 0.55))
    elif int(left_mask.sum()) > 30:
        result.left_line = _fit_line_xz(gx[left_mask], gz[left_mask], config.ransac_thresh_m)
        result.notes.append("left_raw_points")

    if right_env_x.size >= config.min_trunks_for_line:
        result.right_line = _fit_line_xz(right_env_x, right_env_z, max(config.ransac_thresh_m, 0.55))
    elif int(right_mask.sum()) > 30:
        result.right_line = _fit_line_xz(gx[right_mask], gz[right_mask], config.ransac_thresh_m)
        result.notes.append("right_raw_points")

    def trunks_along(line, xs, zs, side):
        if xs.size == 0:
            return
        near = np.abs(line.intercept + line.slope * zs - xs) < 0.50
        for tx, tz, n in _cluster_xz(xs[near], zs[near], config.cluster_eps_m, config.cluster_min_points):
            result.trunks.append(Trunk(side=side, x_m=tx, z_m=tz, n_points=n))

    if result.left_line is not None:
        trunks_along(result.left_line, gx[left_mask], gz[left_mask], "left")
    if result.right_line is not None:
        trunks_along(result.right_line, gx[right_mask], gz[right_mask], "right")

    if result.left_line and result.right_line:
        result.centerline = LineXZ(
            intercept=0.5 * (result.left_line.intercept + result.right_line.intercept),
            slope=0.5 * (result.left_line.slope + result.right_line.slope),
        )
        result.row_width_m = abs(result.right_line.x_at(6.0) - result.left_line.x_at(6.0))
        result.lateral_error_m = -result.centerline.intercept
        result.heading_error_rad = -result.centerline.heading_rad()
        if result.trunks:
            result.depth_range_m = float(max(t.z_m for t in result.trunks))
    elif result.left_line:
        assumed = result.left_line.intercept + config.row_width_m
        result.centerline = LineXZ(
            intercept=0.5 * (result.left_line.intercept + assumed), slope=result.left_line.slope
        )
        result.lateral_error_m = -result.centerline.intercept
        result.heading_error_rad = -result.centerline.heading_rad()
        result.notes.append("right_side_missing_assumed_width")
    elif result.right_line:
        assumed = result.right_line.intercept - config.row_width_m
        result.centerline = LineXZ(
            intercept=0.5 * (result.right_line.intercept + assumed), slope=result.right_line.slope
        )
        result.lateral_error_m = -result.centerline.intercept
        result.heading_error_rad = -result.centerline.heading_rad()
        result.notes.append("left_side_missing_assumed_width")

    vp_rgb = _vanishing_from_rgb(frame.rgb, k.cx)
    if vp_rgb is not None and _plausible_vp(vp_rgb, k.width, k.height):
        result.heading_from_rgb_rad = -float(np.arctan((vp_rgb[0] - k.cx) / k.fx))
        result.vanishing_point_uv = vp_rgb

    if result.left_line and result.right_line:
        p_l = _project_line(result.left_line, 2.0, 28.0, 0.9, frame, config)
        p_r = _project_line(result.right_line, 2.0, 28.0, 0.9, frame, config)
        if p_l and p_r:
            vp_depth = _intersect_image_lines(
                _line_from_two_points(*p_l[0], *p_l[1]),
                _line_from_two_points(*p_r[0], *p_r[1]),
            )
            if vp_depth is not None and np.isfinite(vp_depth).all():
                result.vanishing_point_uv = vp_depth
                if (
                    result.heading_error_rad is not None
                    and result.heading_from_rgb_rad is not None
                    and abs(result.heading_error_rad - result.heading_from_rgb_rad) < np.deg2rad(8.0)
                ):
                    result.heading_error_rad = 0.70 * result.heading_error_rad + 0.30 * result.heading_from_rgb_rad
                    result.notes.append("fused_rgb_heading")

    n_trk = len(result.trunks)
    conf = 0.0
    if result.centerline is not None:
        conf += 0.45
    if n_trk >= 4:
        conf += 0.25
    elif n_trk >= 2:
        conf += 0.12
    if result.left_line and result.right_line:
        conf += 0.20
    if result.vanishing_point_uv is not None:
        conf += 0.10
    result.confidence = float(min(1.0, conf))
    return result

File: cpp/src/perception.cpp
#include "orchard/perception.hpp"
#include "orchard/geometry.hpp"
#include <algorithm>
#include <cmath>
#include <opencv2/imgproc.hpp>
namespace orchard {
namespace {
bool fit_line_xz(const std::vector<float>& xs, const std::vector<float>& zs, float thresh, LineXZ& out) {
    const int n0 = (int)xs.size(); if (n0 < 2) return false;
    std::vector<char> mask(n0, 1); float intercept=0, slope=0;
    for (int iter=0; iter<4; ++iter) {
        double n=0,sumz=0,sumz2=0,sumx=0,sumxz=0;
        for (int i=0;i<n0;++i) if (mask[i]) { n++; sumz+=zs[i]; sumz2+=zs[i]*zs[i]; sumx+=xs[i]; sumxz+=xs[i]*zs[i]; }
        if (n<2) return false;
        double det=n*sumz2-sumz*sumz; if (std::abs(det)<1e-9) return false;
        intercept=(float)((sumz2*sumx-sumz*sumxz)/det); slope=(float)((n*sumxz-sumz*sumx)/det);
        for (int i=0;i<n0;++i) mask[i]=std::abs(intercept+slope*zs[i]-xs[i])<thresh;
    }
    int kept=0; for (char m: mask) kept+=m?1:0; if (kept<2) return false;
    out.intercept=intercept; out.slope=slope; return true;
}
struct Cluster { float x,z; int n; };
std::vector<Cluster> cluster_xz(const std::vector<float>& xs, const std::vector<float>& zs, float eps, int min_pts) {
    const int n=(int)xs.size(); std::vector<Cluster> out; if(!n) return out;
    std::vector<int> order(n); for(int i=0;i<n;++i) order[i]=i;
    std::sort(order.begin(), order.end(), [&](int a,int b){ return zs[a]<zs[b]; });
    std::vector<char> used(n,0); float eps2=eps*eps;
    for (int oi: order) {
        if (used[oi]) continue;
        float sx=0,sz=0; int count=0; std::vector<int> members;
        for (int j=0;j<n;++j) { float dx=xs[j]-xs[oi], dz=zs[j]-zs[oi]; if (dx*dx+dz*dz<eps2) { members.push_back(j); sx+=xs[j]; sz+=zs[j]; count++; } }
        if (count<min_pts) continue;
        for (int j: members) used[j]=1;
        out.push_back({sx/count, sz/count, count});
    }
    return out;
}
void inner_envelope(const std::vector<float>& xs, const std::vector<float>& zs, bool left,
                    std::vector<float>& ox, std::vector<float>& oz, float z_bin=0.9f, int min_pts=8) {
    ox.clear(); oz.clear(); if (xs.empty()) return;
    float z0=zs[0], z1=zs[0]; for (float z: zs) { z0=std::min(z0,z); z1=std::max(z1,z); }
    for (float b0=z0; b0<z1; b0+=z_bin) {
        float b1=b0+z_bin, best=left?-1e9f:1e9f; int count=0;
        for (size_t i=0;i<xs.size();++i) if (zs[i]>=b0 && zs[i]<b1) { count++; best=left?std::max(best,xs[i]):std::min(best,xs[i]); }
        if (count<min_pts) continue; ox.push_back(best); oz.push_back(0.5f*(b0+b1));
    }
}
struct PenCluster { float x,y,z; int n; float xz_span, y_span; };
std::vector<PenCluster> cluster_pens(const std::vector<float>& xs, const std::vector<float>& ys,
                                    const std::vector<float>& zs, float eps, int min_pts) {
    const int n=(int)xs.size(); std::vector<PenCluster> out; if(!n) return out;
    std::vector<int> order(n); for(int i=0;i<n;++i) order[i]=i;
    std::sort(order.begin(), order.end(), [&](int a,int b){ return zs[a]<zs[b]; });
    std::vector<char> used(n,0); float eps2=eps*eps;
    for (int oi: order) {
        if (used[oi]) continue;
        float sx=0,sy=0,sz=0,xmin=1e9f,xmax=-1e9f,ymin=1e9f,ymax=-1e9f,zmin=1e9f,zmax=-1e9f; int count=0;
        std::vector<int> members;
        for (int j=0;j<n;++j) {
            float dx=xs[j]-xs[oi], dz=zs[j]-zs[oi];
            if (dx*dx+dz*dz<eps2) {
                members.push_back(j); sx+=xs[j]; sy+=ys[j]; sz+=zs[j]; count++;
                xmin=std::min(xmin,xs[j]); xmax=std::max(xmax,xs[j]);
                ymin=std::min(ymin,ys[j]); ymax=std::max(ymax,ys[j]);
                zmin=std::min(zmin,zs[j]); zmax=std::max(zmax,zs[j]);
            }
        }
        if (count<min_pts) continue;
        for (int j: members) used[j]=1;
        float span=std::hypot(xmax-xmin, zmax-zmin);
        out.push_back({sx/count, sy/count, sz/count, count, span, ymax-ymin});
    }
    return out;
}
RowPerception perceive_pens(const Frame& frame, const Config& cfg) {
    RowPerception result;
    if (frame.depth_m.empty()) { result.notes.emplace_back("no_depth"); return result; }
    const cv::Mat& depth=frame.depth_m; const Intrinsics& K=frame.K;
    const int stride=std::max(1, cfg.pixel_stride);
    std::vector<float> xs,ys,zs;
    for (int v=0; v<depth.rows; v+=stride) {
        const float* row=depth.ptr<float>(v);
        for (int u=0; u<depth.cols; u+=stride) {
            float z=row[u]; if (z<cfg.desk_depth_min_m || z>cfg.desk_depth_max_m) continue;
            float x,y; K.pixel_to_cam((float)u,(float)v,z,x,y);
            if (std::abs(y)>cfg.desk_y_band_m) continue;
            xs.push_back(x); ys.push_back(y); zs.push_back(z);
        }
    }
    if ((int)xs.size()<20) { result.notes.emplace_back("no_pen_band"); return result; }
    std::vector<PenCluster> left, right;
    for (const auto& c: cluster_pens(xs,ys,zs,cfg.desk_cluster_eps_m,cfg.desk_cluster_min_points)) {
        if (c.xz_span>cfg.desk_max_span_m) continue;
        if (c.y_span<cfg.desk_min_vertical_m && c.n<cfg.desk_cluster_min_points*3) continue;
        if (c.x<-0.02f) left.push_back(c);
        else if (c.x>0.02f) right.push_back(c);
    }
    std::sort(left.begin(), left.end(), [](const PenCluster& a, const PenCluster& b){ return a.z<b.z; });
    std::sort(right.begin(), right.end(), [](const PenCluster& a, const PenCluster& b){ return a.z<b.z; });
    if (left.empty()) result.notes.emplace_back("left_pen_missing");
    if (right.empty()) result.notes.emplace_back("right_pen_missing");
    if (left.empty() || right.empty()) { result.notes.emplace_back("need_both_pens"); return result; }
    const PenCluster& L=left[0]; const PenCluster& R=right[0];
    const float gap=R.x-L.x;
    if (gap<cfg.desk_min_gap_m || gap>cfg.desk_max_gap_m) { result.notes.emplace_back("pen_gap_rejected"); return result; }
    const float center=0.5f*(L.x+R.x); const float lat=-center;
    if (std::abs(lat)>cfg.desk_max_lat_m) { result.notes.emplace_back("lat_rejected"); return result; }
    result.left_line=LineXZ{L.x,0.f}; result.right_line=LineXZ{R.x,0.f}; result.centerline=LineXZ{center,0.f};
    result.lateral_error_m=lat; result.heading_error_rad=0.f; result.row_width_m=gap;
    result.trunks.push_back({Trunk::Left,L.x,L.z,L.n});
    result.trunks.push_back({Trunk::Right,R.x,R.z,R.n});
    result.confidence=std::abs(lat)<0.15f?0.85f:0.70f;
    result.notes.emplace_back("desk_pens");
    return result;
}
}  // namespace
RowPerception perceive(const Frame& frame, const Config& cfg) {
    if (cfg.desk_mode) return perceive_pens(frame, cfg);
    RowPerception result;
    if (frame.depth_m.empty()) { result.notes.emplace_back("no_depth"); return result; }
    const cv::Mat& depth=frame.depth_m; const Intrinsics& K=frame.K;
    const int stride=std::max(1, cfg.pixel_stride);
    std::vector<float> gx, gz; int valid_n=0;
    for (int v=0; v<depth.rows; v+=stride) {
        const float* row=depth.ptr<float>(v);
        for (int u=0; u<depth.cols; u+=stride) {
            float z=row[u]; if (z<cfg.depth_min_m || z>cfg.depth_max_m) continue; ++valid_n;
            float x,y; K.pixel_to_cam((float)u,(float)v,z,x,y);
            float xr,agl,zf; camera_point_to_ground(x,y,z,cfg.pitch_rad(),cfg.camera_height_m,xr,agl,zf);
            if (agl<cfg.trunk_height_min_m || agl>cfg.trunk_height_max_m) continue;
            if (zf<cfg.depth_min_m || zf>cfg.depth_max_m) continue;
            gx.push_back(xr); gz.push_back(zf);
        }
    }
    if (valid_n<80) { result.notes.emplace_back("too_few_depth_pixels"); return result; }
    std::vector<float> lx,lz,rx,rz;
    for (size_t i=0;i<gx.size();++i) {
        if (gx[i]<-cfg.center_keepout_m) { lx.push_back(gx[i]); lz.push_back(gz[i]); }
        else if (gx[i]>cfg.center_keepout_m) { rx.push_back(gx[i]); rz.push_back(gz[i]); }
    }
    std::vector<float> lex,lez,rex,rez;
    inner_envelope(lx,lz,true,lex,lez); inner_envelope(rx,rz,false,rex,rez);
    LineXZ left,right; bool have_l=false, have_r=false;
    if ((int)lex.size()>=cfg.min_trunks_for_line) have_l=fit_line_xz(lex,lez,std::max(cfg.ransac_thresh_m,0.55f),left);
    else if (lx.size()>30) have_l=fit_line_xz(lx,lz,cfg.ransac_thresh_m,left);
    if ((int)rex.size()>=cfg.min_trunks_for_line) have_r=fit_line_xz(rex,rez,std::max(cfg.ransac_thresh_m,0.55f),right);
    else if (rx.size()>30) have_r=fit_line_xz(rx,rz,cfg.ransac_thresh_m,right);
    if (have_l) result.left_line=left; if (have_r) result.right_line=right;
    auto trunks_along=[&](const LineXZ& line, const std::vector<float>& xs, const std::vector<float>& zs, Trunk::Side side){
        std::vector<float> nx,nz;
        for (size_t i=0;i<xs.size();++i) if (std::abs(line.x_at(zs[i])-xs[i])<0.50f) { nx.push_back(xs[i]); nz.push_back(zs[i]); }
        for (const auto& c: cluster_xz(nx,nz,cfg.cluster_eps_m,cfg.cluster_min_points))
            result.trunks.push_back({side,c.x,c.z,c.n});
    };
    if (have_l) trunks_along(left,lx,lz,Trunk::Left);
    if (have_r) trunks_along(right,rx,rz,Trunk::Right);
    if (have_l && have_r) {
        LineXZ c; c.intercept=0.5f*(left.intercept+right.intercept); c.slope=0.5f*(left.slope+right.slope);
        result.centerline=c; result.row_width_m=std::abs(right.x_at(6.f)-left.x_at(6.f));
        result.lateral_error_m=-c.intercept; result.heading_error_rad=-c.heading_rad();
    } else if (have_l) {
        LineXZ c; c.intercept=left.intercept+0.5f*cfg.row_width_m; c.slope=left.slope;
        result.centerline=c; result.lateral_error_m=-c.intercept; result.heading_error_rad=-c.heading_rad();
    } else if (have_r) {
        LineXZ c; c.intercept=right.intercept-0.5f*cfg.row_width_m; c.slope=right.slope;
        result.centerline=c; result.lateral_error_m=-c.intercept; result.heading_error_rad=-c.heading_rad();
    }
    float conf=0; if (result.centerline) conf+=0.45f;
    if (result.trunks.size()>=4) conf+=0.35f; else if (result.trunks.size()>=2) conf+=0.12f;
    if (have_l && have_r) conf+=0.20f;
    result.confidence=std::min(1.f,conf);
    return result;
}
}
